atlasdb-commons/src/main/java/com/palantir/util/IndexEncodingUtils.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.util;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreType;
+import com.google.common.collect.Maps;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.zip.CRC32;
+import org.immutables.value.Value;
+
+public final class IndexEncodingUtils {
+    private IndexEncodingUtils() {}
+
+    /**
+     * Given a set of keys and a map of keys to values, we first order the keys in some fixed manner and then compute
+     * a map that uses indices into this list of keys instead of actual keys.
+     * Values of the returned map are created using the provided mapping function.
+     * We return the list of keys, the index map, and a CRC32 checksum of the ordered keys.
+     *
+     * @param keys a collection of keys (must not be ordered)
+     * @param keyToValue a map of keys to values. Every key MUST be contained in the list of keys.
+     * @param valueMapper a mapping function to run before putting a value into the result map
+     * @param byteMapper a function that extracts bytes from a key object, which are used to compute the checksum
```

Comment:
What restrictions must there be on bytemapper? (hint: determinism, e.t.c)

Reply:
Thinking about this, passing something like a byte mapper can be quite dangerous and present us with the same problems as the checksum (what happens when we want to change the byte mapper?). Also, we have to ensure that we use the same mapper on both ends (client/server). I have removed this argument for now.